File: client/views/register_view.py
from __future__ import annotations

import flet

from api.http_client import APIClient, ConflictError, ValidationError
from localization import t
from state import AppState, UserDTO


def _generate_keypairs():
    """Run in a thread pool — generates both keypairs synchronously."""
    from crypto.key_generator import KeyGenerator

    ed25519_priv, ed25519_pub = KeyGenerator.generate_identity_keypair()
    x25519_priv, x25519_pub = KeyGenerator.generate_prekey_keypair()
    return ed25519_priv, ed25519_pub, x25519_priv, x25519_pub


def register_view(page: flet.Page, state: AppState) -> None:
    page.bgcolor = "#f0f2f5"
    page.overlay.clear()

    username_field = flet.TextField(
        label=t("register.username"),
        autofocus=True,
        bgcolor="#ffffff",
        border_color="#e0e0e0",
        color="#111b21",
    )
    email_field = flet.TextField(
        label=t("register.email"),
        bgcolor="#ffffff",
        border_color="#e0e0e0",
        color="#111b21",
    )
    password_field = flet.TextField(
        label=t("register.password"),
        password=True,
        can_reveal_password=True,
        bgcolor="#ffffff",
        border_color="#e0e0e0",
        color="#111b21",
    )

    username_error = flet.Text("", color="#ea4335", visible=False, size=12)
    email_error = flet.Text("", color="#ea4335", visible=False, size=12)
    password_error = flet.Text("", color="#ea4335", visible=False, size=12)
    general_error = flet.Text("", color="#ea4335", visible=False)

    submit_btn = flet.ElevatedButton(
        t("register.submit"),
        width=300,
        style=flet.ButtonStyle(
            bgcolor="#008069",
            color="#ffffff",
            shape=flet.RoundedRectangleBorder(radius=8),
            padding=flet.padding.symmetric(vertical=16),
        ),
    )
    loading = flet.ProgressRing(visible=False, width=20, height=20, color="#008069")

    def _clear_errors() -> None:
        for txt in (username_error, email_error, password_error, general_error):
            txt.value = ""
            txt.visible = False

    async def do_register(e: flet.ControlEvent) -> None:
        _clear_errors()
        submit_btn.disabled = True
        loading.visible = True
        page.update()

        client = APIClient(state=state)
        try:
            import asyncio
            import base64
            import logging
            from crypto.key_generator import KeyGenerator
            from crypto.key_backup import KeyBackupManager

            logging.info("[Registration] Generating keypairs (thread pool)...")
            # Run blocking key generation off the event loop
            (
                ed25519_priv,
                ed25519_pub,
                x25519_priv,
                x25519_pub,
            ) = await asyncio.to_thread(_generate_keypairs)

            # Serialize public keys
            ed25519_pub_bytes = KeyGenerator.serialize_public_key(ed25519_pub)
            x25519_pub_bytes = KeyGenerator.serialize_public_key(x25519_pub)

            # Create encrypted backup — PBKDF2 runs in thread pool
            logging.info("[Registration] Creating encrypted backup (thread pool)...")
            encrypted_backup = await KeyBackupManager.encrypt_backup_async(
                ed25519_priv,
                x25519_priv,
                password_field.value or "",
            )

            # Base64-encode for transmission
            ed25519_pub_b64 = base64.b64encode(ed25519_pub_bytes).decode("utf-8")
            x25519_pub_b64 = base64.b64encode(x25519_pub_bytes).decode("utf-8")
            encrypted_backup_b64 = base64.b64encode(encrypted_backup).decode("utf-8")

            logging.info("[Registration] Registering user with server...")
            await client.register(
                username=username_field.value or "",
                email=email_field.value or "",
                password=password_field.value or "",
                identity_pub_ed25519=ed25519_pub_b64,
                identity_pub_x25519=x25519_pub_b64,
                encrypted_backup=encrypted_backup_b64,
            )

            # Store private keys in state
            state.ed25519_private = ed25519_priv
            state.x25519_private = x25519_priv
            logging.info("[Registration] Keys stored in state")

            token_data = await client.login(
                username=username_field.value or "",
                password=password_field.value or "",
            )
            state.token = token_data["access_token"]
            me = await client.get_me()
            state.current_user = UserDTO(
                id=me["id"],
                username=me["username"],
                email=me["email"],
                display_name=me.get("display_name"),
            )
            from views.chat_list_view import chat_list_view

            chat_list_view(page, state)
        except Exception as exc:
            # Handle key generation errors
            import logging

            logging.error(f"[Registration] Error: {exc}", exc_info=True)

            msg = str(exc).lower()
            if "cryptographic" in msg or "key generation" in msg:
                general_error.value = t("register.error_crypto", exc=exc)
                general_error.visible = True
            elif isinstance(exc, ConflictError):
                if "username" in msg:
                    username_error.value = exc.message
                    username_error.visible = True
                elif "email" in msg:
                    email_error.value = exc.message
                    email_error.visible = True
                else:
                    general_error.value = exc.message
                    general_error.visible = True
            elif isinstance(exc, ValidationError):
                if "password" in msg:
                    password_error.value = exc.message
                    password_error.visible = True
                elif "email" in msg:
                    email_error.value = exc.message
                    email_error.visible = True
                elif "username" in msg:
                    username_error.value = exc.message
                    username_error.visible = True
                else:
                    general_error.value = exc.message
                    general_error.visible = True
            else:
                general_error.value = t("register.error_server", exc=exc)
                general_error.visible = True
            submit_btn.disabled = False
            loading.visible = False
            page.update()
        finally:
            await client.aclose()

    submit_btn.on_click = do_register
    password_field.on_submit = do_register

    def go_login(e: flet.ControlEvent) -> None:
        from views.login_view import login_view

        login_view(page, state)

    page.controls.clear()
    page.add(
        flet.Column(
            controls=[
                flet.Container(expand=True),
                flet.Card(
                    content=flet.Container(
                        content=flet.Column(
                            controls=[
                                flet.Icon(flet.Icons.CHAT, size=56, color="#008069"),
                                flet.Text(
                                    "Telecommunicator",
                                    size=28,
                                    weight=flet.FontWeight.BOLD,
                                    color="#111b21",
                                ),
                                flet.Text(
                                    t("register.subtitle"), size=14, color="#667781"
                                ),
                                flet.Divider(height=20, color=flet.Colors.TRANSPARENT),
                                username_field,
                                username_error,
                                email_field,
                                email_error,
                                password_field,
                                password_error,
                                general_error,
                                flet.Row(
                                    controls=[submit_btn, loading],
                                    alignment=flet.MainAxisAlignment.CENTER,
                                    vertical_alignment=flet.CrossAxisAlignment.CENTER,
                                ),
                                flet.TextButton(
                                    t("register.have_account"),
                                    on_click=go_login,
                                    style=flet.ButtonStyle(color="#008069"),
                                ),
                            ],
                            alignment=flet.MainAxisAlignment.CENTER,
                            horizontal_alignment=flet.CrossAxisAlignment.CENTER,
                            width=320,
                            spacing=12,
                        ),
                        padding=32,
                    ),
                    elevation=2,
                    bgcolor="#ffffff",
                ),
                flet.Container(expand=True),
            ],
            alignment=flet.MainAxisAlignment.CENTER,
            horizontal_alignment=flet.CrossAxisAlignment.CENTER,
            expand=True,
        )
    )
    page.update()

File: flutter_client/lib/views/register_view.dart
import 'dart:convert';

import 'package:flutter/material.dart';

import '../api/api_client.dart';
import '../config.dart';
import '../crypto/key_backup.dart';
import '../crypto/keys.dart';
import '../l10n/strings.dart';
import '../state/app_state.dart';
import '../ui/theme.dart';
import 'chat_list_view.dart';
import 'login_view.dart';

class RegisterView extends StatefulWidget {
  const RegisterView({super.key, required this.state});
  final AppState state;

  @override
  State<RegisterView> createState() => _RegisterViewState();
}

class _RegisterViewState extends State<RegisterView> {
  final _usernameCtrl = TextEditingController();
  final _emailCtrl = TextEditingController();
  final _passwordCtrl = TextEditingController();
  bool _busy = false;
  String? _usernameError;
  String? _emailError;
  String? _passwordError;
  String? _generalError;

  @override
  void dispose() {
    _usernameCtrl.dispose();
    _emailCtrl.dispose();
    _passwordCtrl.dispose();
    super.dispose();
  }

  void _clearErrors() {
    setState(() {
      _usernameError = null;
      _emailError = null;
      _passwordError = null;
      _generalError = null;
    });
  }

  Future<void> _doRegister() async {
    final state = widget.state;
    _clearErrors();
    setState(() => _busy = true);

    final handle = _usernameCtrl.text.trim();
    final (username, server) = parseHandle(handle);
    if (handle.contains('@') && server == null) {
      setState(() {
        _busy = false;
        _generalError = L10n.t('register.error_username_format');
      });
      return;
    }

    if (server != null) {
      try {
        final (api, ws) = buildApiUrls(server);
        if (api != state.apiUrl) {
          state.apiUrl = api;
          state.wsUrl = ws;
          await state.settings?.set('settings.api_url', api);
          await state.settings?.set('settings.ws_url', ws);
          await closeSharedClients();
        }
      } catch (_) {
        setState(() {
          _busy = false;
          _generalError = L10n.t('register.error_username_format');
        });
        return;
      }
    }

    final client = ApiClient(state: state);
    try {
      final (edPriv, edPub) = await generateEd25519Keypair();
      final (xPriv, xPub) = await generateX25519Keypair();

      final encryptedBackup = await KeyBackupManager.encryptBackup(
        edPriv, xPriv, _passwordCtrl.text);

      await client.register(
        username: username,
        email: _emailCtrl.text,
        password: _passwordCtrl.text,
        identityPubEd25519: base64Encode(edPub),
        identityPubX25519: base64Encode(xPub),
        encryptedBackup: base64Encode(encryptedBackup),
      );

      state.crypto = CryptoKeys(
        ed25519Private: edPriv,
        ed25519Public: edPub,
        x25519Private: xPriv,
        x25519Public: xPub,
      );

      final tokenData = await client.login(username, _passwordCtrl.text);
      state.token = tokenData['access_token'] as String;
      final me = await client.getMe();
      state.currentUser = UserDTO.fromJson(me);

      await state.settings?.set('auth.token', state.token!);
      final settings = state.settings;
      if (settings != null) {
        final supportDir = await settings.supportDir();
        await state.initCryptoStores(supportDir, state.currentUser!.username);
      }

      if (!mounted) return;
      navigateReplacing(context, ChatListView(state: state));
    } on ConflictError catch (e) {
      final msg = e.message.toLowerCase();
      setState(() {
        _busy = false;
        if (msg.contains('username')) {
          _usernameError = e.message;
        } else if (msg.contains('email')) {
          _emailError = e.message;
        } else {
          _generalError = e.message;
        }
      });
    } on ValidationError catch (e) {
      final msg = e.message.toLowerCase();
      setState(() {
        _busy = false;
        if (msg.contains('password')) {
          _passwordError = e.message;
        } else if (msg.contains('email')) {
          _emailError = e.message;
        } else if (msg.contains('username')) {
          _usernameError = e.message;
        } else {
          _generalError = e.message;
        }
      });
    } on Exception catch (e) {
      final s = e.toString();
      setState(() {
        _busy = false;
        _generalError = s.isEmpty
            ? L10n.t('register.error_server', const {})
            : (s.toLowerCase().contains('cryptographic')
                ? L10n.t('register.error_crypto', {'exc': s})
                : L10n.t('register.error_server', {'exc': s}));
      });
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: context.surfaceContainer,
      body: Center(
        child: SingleChildScrollView(
          padding: const EdgeInsets.all(32),
          child: Container(
            width: 360,
            padding: const EdgeInsets.all(32),
            decoration: BoxDecoration(
              color: context.surface,
              borderRadius: BorderRadius.circular(20),
              border: Border.all(color: context.outlineVariant),
              boxShadow: const [
                BoxShadow(
                  blurRadius: 24,
                  spreadRadius: -4,
                  color: Color(0x1A000000),
                  offset: Offset(0, 8),
                ),
              ],
            ),
            child: Column(
              mainAxisSize: MainAxisSize.min,
              children: [
                initialsAvatar(context, 'Telecommunicator', size: 64),
                const SizedBox(height: 10),
                const Text('Telecommunicator',
                    style: TextStyle(fontSize: 26, fontWeight: FontWeight.bold)),
                Text(
                  L10n.t('register.subtitle'),
                  style: TextStyle(fontSize: 14, color: context.onSurfaceVariant),
                ),
                const SizedBox(height: 16),
                TextField(
                  controller: _usernameCtrl,
                  autofocus: true,
                  decoration: themedFieldDecoration(
                    label: L10n.t('register.username'),
                    hintText: L10n.t('register.username_hint'),
                  ),
                ),
                if (_usernameError != null)
                  Text(_usernameError!,
                      style: TextStyle(fontSize: 12, color: context.error)),
                const SizedBox(height: 8),
                TextField(
                  controller: _emailCtrl,
                  decoration: themedFieldDecoration(label: L10n.t('register.email')),
                ),
                if (_emailError != null)
                  Text(_emailError!,
                      style: TextStyle(fontSize: 12, color: context.error)),
                const SizedBox(height: 8),
                TextField(
                  controller: _passwordCtrl,
                  obscureText: true,
                  decoration:
                      themedFieldDecoration(label: L10n.t('register.password')),
                  onSubmitted: (_) => _doRegister(),
                ),
                if (_passwordError != null)
                  Text(_passwordError!,
                      style: TextStyle(fontSize: 12, color: context.error)),
                if (_generalError != null)
                  Padding(
                    padding: const EdgeInsets.only(top: 8),
                    child: Text(_generalError!,
                        style: TextStyle(color: context.error)),
                  ),
                const SizedBox(height: 12),
                Row(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: [
                    primaryButton(
                      context,
                      L10n.t('register.submit'),
                      onPressed: _busy ? null : _doRegister,
                    ),
                    const SizedBox(width: 12),
                    if (_busy)
                      const SizedBox(
                        width: 20,
                        height: 20,
                        child: CircularProgressIndicator(strokeWidth: 2),
                      ),
                  ],
                ),
                TextButton(
                  onPressed: () {
                    Navigator.of(context).pushReplacement(
                      MaterialPageRoute(
                        builder: (_) => LoginView(state: widget.state),
                      ),
                    );
                  },
                  child: Text(L10n.t('register.have_account'),
                      style: TextStyle(color: context.primary)),
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}
